### PR TITLE
PROD-1330: logging for root user access

### PR DESCRIPTION
**Pull request by adamsachs**
Closes https://ethyca.atlassian.net/browse/PROD-1330

### Description Of Changes

To help mitigate and be aware of a potential security weakness in production deployments, we've decided to provide some logging when the root user is in use. Specifically, we're providing:
- a warning message on server startup (both fides and fidesplus)
- a warning message whenever the root user credentials are successfully used for a user "login"
- a warning message whenever the root oauth client credentials are used to successfully acquire a client access token

Below is some logging i captured on my local fides server with representative actions as root user, to get an idea of what logs look like. To me, the only real questions are:
- ~should we change "info message when token is merely requested, warning if the token is actually acquired" to either be warning for both, or just warn when the token is requested?~ edit: answered in comments and adjusted code
- ~is the info message on every root user API call too noisy? i definitely think a warn-level log on every root user API call is too noisy, which is why i went with info for now. but would be curious to hear others' POV.~ edit: answered in comments and adjusted code

-------
Sample logging output on representative actions from local fides server

- on (fides) server startup:
```
fides  | 2023-11-15 13:55:46.954 | DEBUG    | fides.api.util.cache:get_cache:188 - Redis connection succeeded.
fides  | 2023-11-15 13:55:46.954 | DEBUG    | fides.api.app_setup:check_redis:221 - Connection to cache succeeded
fides  | 2023-11-15 13:55:46.956 | INFO     | fides.api.service.privacy_request.email_batch_service:initiate_scheduled_batch_email_send:136 - Initiating scheduler for batch email send
fides  | 2023-11-15 13:55:47.024 | DEBUG    | fides.api.main:setup_server:282 - Sending startup analytics events...
fides  | 2023-11-15 13:55:47.024 | INFO     | fides.api.main:setup_server:296 -
fides  |
fides  | ███████╗██╗██████╗ ███████╗███████╗
fides  | ██╔════╝██║██╔══██╗██╔════╝██╔════╝
fides  | █████╗  ██║██║  ██║█████╗  ███████╗
fides  | ██╔══╝  ██║██║  ██║██╔══╝  ╚════██║
fides  | ██║     ██║██████╔╝███████╗███████║
fides  | ╚═╝     ╚═╝╚═════╝ ╚══════╝╚══════╝
fides  |
fides  | 2023-11-15 13:55:47.025 | WARNING  | fides.api.main:warn_root_user_enabled:252 - Security warning: you are running Fides with a root user configured! This is not recommended in production deployments
fides  | 2023-11-15 13:55:47.025 | INFO     | fides.api.main:setup_server:300 - Fides startup complete! v2.24.1b0.post0.dev6
fides  | 2023-11-15 13:55:47.025 | INFO     | fides.api.main:setup_server:302 - Server setup completed in 4.628 seconds
```

- on root user login from admin UI:
```
fides  | 2023-11-17 14:46:04.378 | INFO     | fides.api.main:log_request:166 - Request received | {'method': 'POST', 'status_code': 403, 'handler_time': '22.441ms', 'path': '/api/v1/login/'}
fides  | 2023-11-17 14:46:18.070 | INFO     | fides.api.main:log_request:166 - Request received | {'method': 'GET', 'status_code': 200, 'handler_time': '46.585ms', 'path': '/health'}
fides  | 2023-11-17 14:46:22.383 | WARNING  | fides.api.api.v1.endpoints.user_endpoints:user_login:553 - Successful root user login!
fides  | 2023-11-17 14:46:22.384 | INFO     | fides.api.api.v1.endpoints.user_endpoints:user_login:583 - Creating login access token
fides  | 2023-11-17 14:46:22.391 | INFO     | fides.api.main:log_request:166 - Request received | {'method': 'POST', 'status_code': 200, 'handler_time': '15.822ms', 'path': '/api/v1/login/'}
```

- root client authentication from swagger API UI:
```
fides  | 2023-11-15 14:08:00.977 | INFO     | fides.api.api.v1.endpoints.oauth_endpoints:acquire_access_token:105 - Creating access token
fides  | 2023-11-15 14:08:00.982 | WARNING  | fides.api.api.v1.endpoints.oauth_endpoints:acquire_access_token:112 - Root user access token acquired!
fides  | 2023-11-15 14:08:00.987 | INFO     | fides.api.main:log_request:166 - Request received | {'method': 'POST', 'status_code': 200, 'handler_time': '560.952ms', 'path': '/api/v1/oauth/token'}
```

### Code Changes

* [x] log messaging described above

### Steps to Confirm

* [x] ran fides locally and performed root user actions, checked logs
* [ ] should confirm with fidesplus that the same logging occurs (likely want to use an alpha tag)

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### CODE REVIEW

**Comment by adamsachs** [comment]
ok @daveqnet and @NevilleS i think this is ready for some review! we can easily adjust the logging as you see fit, i tried to give a good rundown of what this first attempt looks like.

**Comment by daveqnet** [comment]
Testing this PR branch from a black box perspective, running `nox -s "fides_env(test)" on the following commit:

```
~/projects/github/ethyca/fides asachs/PROD-1330 ❯ git log -1 
commit 9c332a446f53b3d9919c9eaaea004dfc7c4e1562 (HEAD -> asachs/PROD-1330, tag: 2.24.1a0, origin/asachs/PROD-1330)
Author: Adam Sachs <adam@ethyca.com>
Date:   Wed Nov 15 10:02:15 2023 -0500

    add missing return type annotation
```

#### 1. WARN log on start-up ✅ 

```
2023-11-16 16:58:54 2023-11-16 16:58:54.742 | WARNING  |  - Security warning: you are running Fides with a root user configured! This is not recommended in production deployments
```

#### 2. WARN log on `root_user` login ✅ 

```
2023-11-16 17:07:28 2023-11-16 17:07:28.524 | WARNING  |  - Performing root user login!
```

#### 3. WARN log on `oauth ✅ 

```
2023-11-16 17:10:56 2023-11-16 17:10:56.283 | WARNING  |  - Root user access token acquired!
```

So my high-level security requirements are met, thanks @adamsachs! This is just a nit so feel free to ignore, is it possible to change the `Performing root user login!` text. "Performing" implies that it hasn't succeeded yet, and my ask is for a log when it succeeds.

You could argue that failed login attempts are also important, but I'm happy to keep that out of scope for this ticket.

Re your questions:

> should we change "info message when token is merely requested, warning if the token is actually acquired" to either be warning for both, or just warn when the token is requested?

I'm not sure I understand. Do you basically mean that although there are two ways to get a bearer token with "root" permissions, the point in the code where the token is generated (and could be logged) is the same?

The two methods I mean (with non-sensitive local test creds in both cases):

1. `POST /api/v1/login` with `root_username/root_password` credentials in the json body

```http
POST /api/v1/login HTTP/1.1
Host: localhost:8080
Content-Length: 58
sec-ch-ua: "Chromium";v="119", "Not?A_Brand";v="24"
unescape-safestr: true
sec-ch-ua-mobile: ?0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.6045.123 Safari/537.36
content-type: application/json
access-control-allow-origin: *
x-fides-source: fidesops-admin-ui
sec-ch-ua-platform: "macOS"
Accept: */*
Origin: http://localhost:8080
Sec-Fetch-Site: same-origin
Sec-Fetch-Mode: cors
Sec-Fetch-Dest: empty
Referer: http://localhost:8080/login
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
Connection: close

{"username":"root_user","password":"VGVzdHBhc3N3b3JkMSE="}
```

2. `POST /api/v1/oauth/token` with `oauth_root_client_id/oauth_root_client_secret` credentials base64 encoded in the `Authorization` header

```http
POST /api/v1/oauth/token HTTP/1.1
Host: localhost:8080
Content-Length: 29
sec-ch-ua: "Chromium";v="119", "Not?A_Brand";v="24"
sec-ch-ua-mobile: ?0
Authorization: Basic ZmlkZXNhZG1pbjpmaWRlc2FkbWluc2VjcmV0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.6045.123 Safari/537.36
Content-Type: application/x-www-form-urlencoded
Accept: application/json, text/plain, */*
X-Requested-With: XMLHttpRequest
sec-ch-ua-platform: "macOS"
Origin: http://localhost:8080
Sec-Fetch-Site: same-origin
Sec-Fetch-Mode: cors
Sec-Fetch-Dest: empty
Referer: http://localhost:8080/docs
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
Connection: close

grant_type=client_credentials
```

> is the info message on every root user API call too noisy? i definitely think a warn-level log on every root user API call is too noisy, which is why i went with info for now. but would be curious to hear others' POV.

It's not a requirement from me. Again, having audit logging for all user activity (not just root_user activity) would be a useful feature, but it's out of scope for this ticket imo.

**Comment by adamsachs** [comment]
thanks for the testing and feedback @daveqnet !


> > should we change "info message when token is merely requested, warning if the token is actually acquired" to either be warning for both, or just warn when the token is requested?
> 
> I'm not sure I understand. Do you basically mean that although there are two ways to get a bearer token with "root" permissions, the point in the code where the token is generated (and could be logged) is the same?

nope sorry i wasn't clear, it's a simpler question i had - i'm basically referring to my proposed changes [here](https://github.com/ethyca/fides/pull/4432/files#diff-e29be78d90aaa0c5633f9cd3c4846f4e87c38076c55bb10260382e34d4ca3440R86-R113), where i'm logging both the "request" for a token and the actual issuing of the token if the request is successful.

based on your other comments , namely 
> This is just a nit so feel free to ignore, is it possible to change the Performing root user login! text. "Performing" implies that it hasn't succeeded yet, and my ask is for a log when it succeeds.

it seems like we just want to log when root credentials are successfully retrieved, we don't need the logging for each _request_. that's totally fine by me and i'll make some small tweaks, and i think this should be ready to go!

**Comment by daveqnet** [comment]
Perfect, thanks @adamsachs! 🥳

**Comment by adamsachs** [comment]
> Perfect, thanks @adamsachs! 🥳

ok @daveqnet, adjusted and should be good to go! i updated some of the comments to indicate how the logging looks now. i'll wait for CI to pass and also will need an approval :) but then we should be good to go. feel free to run the same tests you did locally to confirm, but i also don't think it's strictly necessary given how minor the tweaks are 👍

**Review by daveqnet — COMMENTED** [review]
Approved from a security / black-box perspective, but please get code review and approveal as well - thanks!

File: src/fides/api/api/v1/endpoints/oauth_endpoints.py
Comment (NevilleS) [inline src/fides/api/api/v1/endpoints/oauth_endpoints.py:108]:
```suggestion
        logger.warning("OAuth Root Client ID was used to generate an API access token. If unexpected, review security settings (FIDES__SECURITY__OAUTH_ROOT_CLIENT_ID)")
```

File: src/fides/api/api/v1/endpoints/user_endpoints.py
Comment (NevilleS) [inline src/fides/api/api/v1/endpoints/user_endpoints.py:553]:
```suggestion
        logger.warning("Root Username & Password were used to login! If unexpected, review security settings (FIDES__SECURITY__ROOT_USERNAME and FIDES__SECURITY__ROOT_PASSWORD)")
```

File: src/fides/api/main.py
Comment (NevilleS) [inline src/fides/api/main.py:254]:
```suggestion
        logger.warning(
            "Root Username & Password are configured and can be used to login as a root user. If unexpected, review security settings (FIDES__SECURITY__ROOT_USERNAME and FIDES__SECURITY__ROOT_PASSWORD)"
        )
```

**Review by NevilleS — APPROVED** [review]
Couple nitpicky log changes 👍

File: src/fides/api/api/v1/endpoints/user_endpoints.py
Comment (adamsachs) [reply src/fides/api/api/v1/endpoints/user_endpoints.py:553]: hmm, this doesn't feel a bit noisy for every root user login?

Comment (adamsachs) [reply src/fides/api/api/v1/endpoints/user_endpoints.py:553]: ok, understood now that we want to provide extra context to prevent mis-reading the log as a concern in cases when it's totally expected 👍

**Pull request merged**
